src/lib/watch-processor.ts

```diff
@@ -100,7 +101,20 @@ async function runBinding(binding: Binding, capabilityNamespaces: string[]) {
         // Then, check if the object matches the filter
         const filterMatch = filterNoMatchReason(binding, obj, capabilityNamespaces);
         if (filterMatch === "") {
-          await binding.watchCallback?.(obj, phase);
+          if (binding.isFinalize) {
+            if (!obj.metadata?.deletionTimestamp) {
+              return;
+            }
+            try {
+              await binding.finalizeCallback?.(obj);
+
+              // irrespective of callback success / failure, remove pepr finalizer
```

Comment:
I can see some use cases where a retry might be preferred. But perhaps that's not the common case.

I'd at least suggesting logging when the `finalizeCallback` errors.

Reply:
As for potential retry logic, agreed -- we think there're probably situations where it would be useful too, just didn't want to crowbar that in without proper research & a specific, user-requested use case -- partly to keep this feature as simple as possible, partly to be sure we weren't assuming what sorts of retry logic would be desired.

I wouldn't be surprised to see a feature like that roll out in the future though, especially once Finalize() sees some use & we get a feel for it's rough edges.